README.md
![](doc/header-clip.gif)
# sync.soccer ![badge-experimental] ![badge-travis]

sync.soccer synchronises (or _aligns_) soccer
datasets, so that an event in one dataset is matched to the corresponding
event or snapshot in the other. This repository contains an 
implementation that aligns Opta's 
(now STATS Perform's) F24 feeds to ChyronHego's Tracab files. More
formats may be added in the future. The heart of sync.soccer is the
[Needleman-Wunsch][link-nwwiki] dynamic programming algorithm. For a 
fuller explanation of the method see separate [blogpost][link-blog]. 

## Table of contents
### [Installation](#installation)<br>
### [Usage](#usage)
### [Known issues and troubleshooting](#troubleshooting)
* [Missing Tracab metadata](#missing-metadata)
* [Performace and system requirements](#performance)
* [Unsatisfactory alignments](#bad-alignments)
### [License](#license)

<a name="installation"/>

## Installation

sync.soccer has been developed and tested on Ubuntu Linux.
Feedback on compilation and performance on other platforms is 
welcome. To build sync.soccer, you need [Stack][link-stack]. 
Then, to download and compile, say:

```sh
$ git clone git@github.com:huffyhenry/sync.soccer.git
$ cd sync.soccer
$ stack build
```
<a name="usage"/>

## Usage

If the build was successful, you can run sync.soccer like so
(note the double dash, which tells Stack that all arguments that follow
are meant for the sync-soccer executable and not for Stack itself):
```
$ stack exec -- sync-soccer --help

Usage: sync-soccer TRACAB-META TRACAB-DATA F7 F24 OUTPUT [-t|--time-only] 
                   [-u|--timestamp] [-s|--show-sync] [-c|--scale-clock X] 
                   [-l|--scale-location X] [-p|--scale-player X] 
                   [-b|--scale-ball X] [-e|--event-csv FILEPATH] 
                   [-f|--frame-csv FILEPATH]
  Synchronise Tracab and F24 data.

Available options:
  -t,--time-only           Sync only by time
  -u,--timestamp           Use F24 event timestamp instead of min:sec
  -s,--show-sync           Print human-readable sync on screen
  -c,--scale-clock X       Clock mismatch leading to unit penalty [s, def=1]
  -l,--scale-location X    Location mismatch leading to unit penalty [m, def=5]
  -p,--scale-player X      Player-ball gap leading to unit penalty [m, def=1]
  -b,--scale-ball X        Penalty for syncing to dead-ball frame [def=5]
  -e,--event-csv FILEPATH  Location to save F24 events CSV
  -f,--frame-csv FILEPATH  Location to save Tracab frames CSV
  -h,--help                Show this help text
```
You must specify five file paths: to Tracab metadata, 
Tracab payload, F7 file (essentially F24 metadata), F24 file and
a path to the output file where the completed alignment will be saved
as a CSV of event and frame IDs.

The meaning of the optional arguments is as follows:
* `--time-only` discards all the added value of the algorithm and matches
  the events to the frame nearest to it in time. Useful for debugging,
  comparisons or when the algorithm misbehaves on a particuar game.
* `--timestamp` uses the F24 event timestamp instead of minute and second
  as the time of the event.
* `--show-sync` will print the alignment on screen. It will look something
  like this:
  ```
  Event <event id> (67:12 Pass)          Frame <frame id> (implied clock: 67:12.920)
  --gap length 55--                            
  Event <event id> (67:15 Pass)          Frame <frame id> (implied clock: 67:15.160)
  --gap length 30--                            
  Event <event id> (67:16 Clearance)     Frame <frame id> (implied clock: 67:16.400)
  --gap length 26--                            
  Event <event id> (67:17 Ball recovery) Frame <frame id> (implied clock: 67:17.480)
  Event <event id> (67:18 Pass)          Frame <frame id> (implied clock: 67:17.520)
  --gap length 35--                            
  Event <event id> (67:19 Pass)          Frame <frame id> (implied clock: 67:18.960)
  --gap length 50--                            
  Event <event id> (67:21 Pass)          Frame <frame id> (implied clock: 67:21.000)
  --gap length 118--                           
  Event <event id> (67:26 Pass)          Frame <frame id> (implied clock: 67:25.760)
  --gap length 104--                           
  Event <event id> (67:30 Pass)          Frame <frame id> (implied clock: 67:29.960)
  --gap length 42--                            
  Event <event id> (67:32 Miss)          Frame <frame id> (implied clock: 67:31.680)
  ```
* The parameters starting with `--scale` control the _relative_ importance of the
  four components of the event-frame similarity score: difference of time, location,
  distance between the player identified by the event and the ball, and ball
  status (alive/dead). The default values were chosen by trial and error, and
  force the algorithm to penalise equally a 5 metre discrepancy in location, 
  1 second discrepancy in time, and 1 metre distance between the active player
  and the ball. Because the similarity score between any event and any frame
  falls in (0, 1), the default `--scale-ball` value of 5 means that the algorithm
  can never align to a dead-ball frame.
* `--event-csv` and `--frame-csv` write the event and frame data to files
  in a format compatible with the visualisation scripts in `viz/`.
  

<a name="troubleshooting"/>

## Known issues and troubleshooting

<a name="missing-metadata"/>

### Missing Tracab metadata
You may have received a Tracab file without the corresponding metadata.
Tracab metadata is a small XML file that nevertheless contains
essential information for the synchronisation algorithm: pitch size, 
frame rate and the beginning and end of the first and second half. 
If you know 
these parameters or can deduce them from raw data, you can mock up a 
metadata file based on this template:

```xml
<TracabMetaData sVersion="1.0">
    <match iId="0" dtDate="0" iFrameRateFps="25" fPitchXSizeMeters="105.00" fPitchYSizeMeters="68.00" fTrackingAreaXSizeMeters="0.00" fTrackingAreaYSizeMeters="0.00">
        <period iId="1" iStartFrame="<value>" iEndFrame="<value>"/>
        <period iId="2" iStartFrame="<value>" iEndFrame="<value>"/>
        <period iId="3" iStartFrame="0" iEndFrame="0"/>
        <period iId="4" iStartFrame="0" iEndFrame="0"/>
    </match>
</TracabMetaData>
```
Currently, sync.soccer cares only about the `iFrameRateFps`, 
`iStartFrame`, `iEndFrame`, `fPitchXSizeMeters` and `fPitchYSizeMeters` 
values. The rest can be set to 0. In fact, you only _really_ need
the `iStartFrame` values: `iEndFrame` can safely be set to something
corresponding to a time after the half ended (at a small performace cost), 
frame rate is usually (always?) 25, and you may get decent enough results 
using the standard pitch size of 105x68 metres (and perhaps increasing the 
`--scale-location` value).

<a name="performance"/>

### Performance and system requirements
The core algorithm needs to build two ~75'000 (# of Tracab frames) x 
~800 (# of F24 events) matrices, one for each half of the match, 
to find the optimal alignment. This requires time and space.
On my machine, the program needs 5-10 minutes and ~10 GB of RAM to
synchronise a single game. There is a lot of low-hanging fruit to be
picked in terms of optimisation, and I expect future versions of the
software to be much more streamlined. Parallel synchronisation of 
game halves is also in the works.

<a name="bad-alignments"/>

### Unsatisfactory alignments
Sometimes an alignment produced by sync.soccer contains segments
that are demonstrably worse than if synchronisation were performed
purely on the basis of time. They are typically a result of discrepancies 
between the F24 and Tracab data that are greater than is reasonable,
for example due to very inaccurate F24 event location or even
Tracab ball tracking error. In such situations, the algorithm may align
event(s) to frame(s) that are a significant time apart, but where 
(by chance) the
locations agree better. Such mis-alignment can then have a knock-on
effect on the next several seconds of data, until everything straightens 
out.

One solution in these situations is to increase the `--scale-location`
parameter, making the algorithm less sensitive to differences in
event and ball locations. Another is to run a second synchronisation
with `--time-only` and manually (or using a script) merge the two sync 
files so that the
problematic events are aligned only based on time. The time slippage
plot in `viz/slippage.R` can help identify the 
potentially problematic segments, which can then be animated with
`viz/animate.R` to see what is going on.

Such pathological situations need to be distinguished from those
when the algorithm produces an imperfect but reasonable
alignment. These may eventually be corrected by developing smarter,
more granular event-frame agreement scoring functions.

<a name="license"/>

## License
Copyright (c) 2018-2020 Allan Clark and Marek Kwiatkowski.

This program is free software: you can redistribute it and/or modify 
it under the terms of the GNU Affero General Public License as published 
by the Free Software Foundation, either version 3 of the License, 
or (at your option) any later version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. 
See the GNU Affero General Public License for more details.

You should have received a copy of the GNU Affero General Public License along with this program. If not, see <https://www.gnu.org/licenses/>. 


[header-clip]: doc/header-clip.gif
[badge-experimental]: https://img.shields.io/badge/lifecycle-proof%20of%20concept-important
[badge-travis]: https://api.travis-ci.org/huffyhenry/sync.soccer.svg?branch=master
[link-nwwiki]: https://en.wikipedia.org/wiki/Needleman%E2%80%93Wunsch_algorithm
[link-blog]: https://kwiatkowski.io/sync.soccer
[link-stack]: https://docs.haskellstack.org/en/stable/README/

CHANGELOG.md
# Revision history for sync.soccer

## 0.1.0.0 -- 2020-07-18

* First version.


src/F24.hs
module F24 where

import qualified Data.Map.Strict as Map
import qualified Tracab as Tcb
import Prelude hiding (min)
import qualified Data.ByteString as BS
import Text.XML.Light.Types (Element)
import Text.Printf (printf)
import Control.Monad (liftM)
import Data.Time
import Data.Time.LocalTime
import Data.Time.Clock (nominalDiffTimeToSeconds)
import Data.Maybe
import XmlUtils (attrLookupStrict, attrLookup, hasAttributeWithValue)
import qualified XmlUtils as Xml

data Game coordinates = Game {
    gid :: Int,
    away_team_id :: Int,
    away_team_name :: String,
    competition_id :: Int,
    competition_name :: String,
    game_date :: LocalTime,
    home_team_id :: Int,
    home_team_name :: String,
    matchday :: Int,
    period_1_start :: LocalTime,
    period_2_start :: LocalTime,
    season_id :: Int,
    season_name :: String,
    events :: [Event coordinates]
}

instance Eq (Game a) where
    g1 == g2 = gid g1 == gid g2

instance Show (Game a) where
    show g = show (gid g) ++ " " ++ home_team_name g ++ " vs " ++ away_team_name g ++ " on " ++ show (game_date g)


type PlayerId = Int

data Event coordinates = Event {
    eid :: Int,
    event_id :: Int,
    type_id :: Int,
    period_id :: Int,
    min :: Int,
    sec :: Int,
    player_id :: Maybe PlayerId,
    team_id :: Int,
    outcome :: Maybe Int,
    coordinates :: Maybe coordinates,
    timestamp :: LocalTime,
    last_modified :: LocalTime,
    qs :: [Q]
}

instance Eq (Event a) where
    e1 == e2 = eid e1 == eid e2

instance Show (Event a) where
    show e = let typeName = eventTypeName e
             in printf "Event %d (%02d:%02d %s)" (eid e) (min e) (sec e) typeName

hasq :: Int -> Event a -> Bool
hasq i e = any (hasQid i) (qs e)

-- FIXME: Consider throwing an exception on duplicated qualifiers
qval :: Int -> Event a -> Maybe String
qval i e = let qq = filter (hasQid i) (qs e)
           in case qq of
                  [] -> Nothing
                  q:_ -> value q

hasQid :: Int -> Q -> Bool
hasQid i q = qualifier_id q == i

-- The amount of seconds played until the event, treating all completed
-- game periods as having lasted exactly 45 minutes.
-- The first argument controls whether event timestamp should be used.
eventClock :: Bool -> Game cs -> Event cs -> Double
eventClock False _ ee = fromIntegral (60 * min ee + sec ee) + 0.5
eventClock True gg ee = nominalOffset + actualSincePeriodStart where
    half = period_id ee
    nominalOffset = fromIntegral (60 * 45 * (half - 1))
    eventTimestamp = timestamp ee
    periodStart = if half == 1 then period_1_start gg else period_2_start gg
    actualSincePeriodStart = convert (diffLocalTime eventTimestamp periodStart)
    convert = read . show . nominalDiffTimeToSeconds


data Q = Q {
    qid :: Int,
    qualifier_id :: Int,
    value :: Maybe String
}

instance Eq Q where
    q1 == q2 = qid q1 == qid q2

instance Show Q where
    show q = "Q" ++ show (qualifier_id q) ++ if isNothing (value q) then "" else " (value: " ++ show (value q) ++ ")"


-- F24 data may be loaded from different resources
class F24Loader a where
    loadGame :: a -> IO (Game F24Coordinates)


data F24Coordinates = F24Coordinates {
    xPercentage :: Float,
    yPercentage :: Float
}

-- Parse date-times such as 2018-09-15T14:56:30.
parseDatetime :: String -> Maybe LocalTime
parseDatetime = parseTimeM True defaultTimeLocale "%Y-%m-%dT%H:%M:%S"

-- Parse timestamps such 2018-09-15T14:19:24.588.
parseTimestamp :: String -> Maybe LocalTime
parseTimestamp = parseTimeM True defaultTimeLocale "%Y-%m-%dT%H:%M:%S%Q"

loadGameFromFile :: String -> IO (Game F24Coordinates)
loadGameFromFile filepath = do
    root <- Xml.loadXmlFromFile filepath
    xml <- BS.readFile filepath
    return (makeGame (head $ Xml.getAllChildren root))


makeQ :: Element -> Q
makeQ el = Q { qid = attrLookupStrict el read "id",
               qualifier_id = attrLookupStrict el read "qualifier_id",
               value = attrLookup el read "value" }

makeEvent :: Element -> Event F24Coordinates
makeEvent el =
    Event
        { eid = attrLookupStrict el read "id",
            event_id = attrLookupStrict el read "event_id",
            type_id = attrLookupStrict el read "type_id",
            period_id = attrLookupStrict el read "period_id",
            min = attrLookupStrict el read "min",
            sec = attrLookupStrict el read "sec",
            player_id = attrLookup el read "player_id",
            team_id = attrLookupStrict el read "team_id",
            outcome = attrLookup el read "outcome",
            coordinates = coordinates,
            timestamp = attrLookupStrict el (fromJust . parseTimestamp) "timestamp",
            last_modified = attrLookupStrict el (fromJust . parseDatetime) "last_modified",
            qs = map makeQ $ Xml.getChildrenWithQName "Q" el
            }
    where
    coordinates =
        do x <- attrLookup el read "x"
           y <- attrLookup el read "y"
           return $ F24Coordinates { xPercentage = x, yPercentage = y }

makeGame :: Element -> Game F24Coordinates
makeGame el = Game { gid = attrLookupStrict el read "id",
                     away_team_id = attrLookupStrict el read "away_team_id",
                     away_team_name = attrLookupStrict el id "away_team_name",
                     competition_id = attrLookupStrict el read "competition_id",
                     competition_name = attrLookupStrict el id "competition_name",
                     game_date = attrLookupStrict el (fromJust . parseDatetime) "game_date",
                     home_team_id = attrLookupStrict el read "home_team_id",
                     home_team_name = attrLookupStrict el id "home_team_name",
                     matchday = attrLookupStrict el read "matchday",
                     period_1_start = attrLookupStrict el (fromJust . parseDatetime) "period_1_start",
                     period_2_start = attrLookupStrict el (fromJust . parseDatetime) "period_2_start",
                     season_id = attrLookupStrict el read "season_id",
                     season_name = attrLookupStrict el id "season_name",
                     events = map makeEvent $ Xml.getChildrenWithQName "Event" el
                   }


isAwayTeam :: Game a -> Event b -> Bool
isAwayTeam game event = team_id event == away_team_id game

isHomeTeam :: Game a -> Event b -> Bool
isHomeTeam game event = team_id event == home_team_id game

eventTeam :: Game a -> Event b -> Maybe Tcb.TeamKind
eventTeam game event
    | isHomeTeam game event = Just Tcb.Home
    | isAwayTeam game event = Just Tcb.Away
    | otherwise = Nothing


getFlippedHalves :: Tcb.Metadata -> Tcb.Frames Tcb.Positions -> (Maybe Tcb.TeamKind, Maybe Tcb.TeamKind)
getFlippedHalves metaData frames =
    (Map.lookup 1 flippedMap, Map.lookup 2 flippedMap)
    where
    flippedMap = createFlippedTeamMapping metaData frames


createFlippedTeamMapping :: Tcb.Metadata -> Tcb.Frames Tcb.Positions -> Map.Map Int Tcb.TeamKind
createFlippedTeamMapping metaData frames =
    Map.fromList $ map createKeyFlipped tracabPeriods
    where
    tracabPeriods = filter (\p -> Tcb.startFrame p /= Tcb.endFrame p) (Tcb.periods metaData)

    createKeyFlipped period =
        ( Tcb.periodId period, flipped)
        where
        flipped =
            case filter isKickOff frames of
                [] ->
                    -- This obviously shouldn't happen but the map is strict and you
                    -- may have incomplete data, for example you may be only analysing
                    -- the second half, in which case you don't care about the first half
                    -- and perhaps haven't provided the frames for the first half.
                    Tcb.Away
                kickOffFrame : _ ->
                    Tcb.rightToLeftKickOff kickOffFrame
        kickOffFrameId = Tcb.startFrame period
        isKickOff frame = Tcb.frameId frame == kickOffFrameId


convertGameCoordinates :: Tcb.Metadata -> Tcb.Frames Tcb.Positions -> Game F24Coordinates -> Game Tcb.Coordinates
convertGameCoordinates metaData frames game =
    game { events = map convertEvent (events game) }
    where
    flippedMap = createFlippedTeamMapping metaData frames
    convertEvent event =
        event { coordinates = fmap convertCoordinates $ coordinates event }
        where
        convertCoordinates coords =
            Tcb.Coordinates
                { Tcb.x = convertedX
                , Tcb.y = convertedY
                }
            where
            isInPenaltyBox =
                (optaX <= 17.0 || optaX >= 83) &&
                (21.1 <= optaY && optaY <= 78.9)

            convertedX = convertCentiMeters pitchLength xCentiMeters
            pitchLength = Tcb.pitchSizeX metaData
            optaX = xPercentage coords
            xCentiMeters =
                -- So note, for the x coordinate to be treated as in the penalty box
                -- the event has to be entirely within a penalty box with respect to
                -- both x and y coordinates, not just the x. Same is true below for y
                -- (which is less controversial).
                case isInPenaltyBox of
                    False ->
                        -- The tracab pitch length is given in *meters* but the x-coordinate of
                        -- a frame is given in centi-meters. The optaX is given in percentage so normally
                        -- to find out how far along the pitch we'd find out what 1% of the pitch is and
                        -- multiply it by the optaX (a percentage). However 1% is just the pitchLength divided
                        -- by 100, but since the pitch length is given in meters it's already divided by 100.
                        optaX * pitchLength
                    True ->
                        case optaX <= 17.0 of
                            True ->
                                yards * 91.44
                                where
                                yards = optaX * 18.0 / 17.0
                            False ->
                                100 * meters
                                where
                                meters = pitchLength - metersFromOppGoal
                                metersFromOppGoal = 0.9144 * yardsFromOppGoal
                                yardsFromOppGoal = percentageFromOppGoal * (18.0 / 17.0)
                                percentageFromOppGoal = 100.0 - optaX

            convertedY = convertCentiMeters pitchWidth yCentiMeters
            pitchWidth = Tcb.pitchSizeY metaData
            optaY = yPercentage coords
            yCentiMeters =
                case isInPenaltyBox of
                    False ->
                        -- See above, basically the pitchWidth is given in meters but the y-coordinate we
                        -- want is in centi-meters.
                        optaY * pitchWidth
                    True ->
                        -- A penalty box extends 18y from both sides of the goal which is itself 8 y.
                        -- For opta the 'bottom' is at 21.1 and the 'top' at 78.9.
                        yards * 91.44
                        where
                        yards = optaY * (18.0 + 18.0 + 8) / (78.9 - 21.1)

            convertCentiMeters pitchSize centimeters =
                -- Again the pitch size is given in meters so we have to multiply by 100.
                perhapsFlipFactor * (round $ centimeters - (pitchSize * 100 / 2.0))


        perhapsFlipFactor =
            case Map.lookup (period_id event) flippedMap of
                Just Tcb.Home | isHomeTeam game event -> -1
                Just Tcb.Away | isAwayTeam game event -> -1
                _                                     -> 1

-- Whether the event is an on-the-ball event.
-- Consult the list below for the meaning of event type IDs.
isOTB :: Event c -> Bool
isOTB e = type_id e `elem` ([1..4] ++ [7..16] ++ [41, 42, 44, 45] ++ [49..61])


eventTypeName :: Event a -> String
eventTypeName event =
    case type_id event of
        1 -> "Pass"
        2 -> "Offside pass"
        3 -> "Take on"
        4 -> "Foul"
        5 -> "Ball out"
        6 -> "Corner awarded"
        7 -> "Tackle"
        8 -> "Interception"
        9 -> "Turnover"
        10 -> "Save"
        11 -> "Claim"
        12 -> "Clearance"
        13 -> "Miss"
        14 -> "Woodwork"
        15 -> "Attempt saved"
        16 -> "Goal"
        17 -> "Card"
        18 -> "Sub off"
        19 -> "Sub on"
        20 -> "Player retires"
        21 -> "Player returns"
        22 -> "Player becomes GK"
        23 -> "GK becomes player"
        24 -> "Conditions change"
        25 -> "Official change"
        -- 26 missing in documentation ditto the others that are commented out
        27 -> "Start delay"
        28 -> "End delay"
        -- 29
        30 -> "Period end"
        -- 31
        32 -> "Period start"
        -- 33
        34 -> "Team setup"
        35 -> "Player changes position"
        36 -> "Player changes jersey number"
        37 -> "Match end"
        38 -> "Temp goal"
        39 -> "Temp attempt"
        40 -> "Formation change"
        41 -> "GK ball punch"
        42 -> "Good skill"
        43 -> "Deleted event"
        44 -> "Aerial"
        45 -> "Failed challenge"
        -- 46
        47 -> "Card rescinded"
        -- 48
        49 -> "Ball recovery"
        50 -> "Dispossesed"
        51 -> "Error"
        52 -> "Keeper pick up"
        53 -> "Cross not claimed"
        54 -> "Smother"
        55 -> "Offside provoked"
        56 -> "Shield ball out"
        57 -> "Foul throw"
        58 -> "Penalty faced"
        59 -> "Keeper sweeper"
        60 -> "Chance missed"
        61 -> "Bad touch"
        -- 62
        63 -> "Temp save"
        64 -> "Resume"
        65 -> "Contentious decision"

        unknown -> "Unknown: " ++ show unknown


type ShirtNumber = Int
type ShirtNumbers = Map.Map PlayerId (Tcb.TeamKind, ShirtNumber)


data Metadata = Metadata{
    homeTeam :: TeamData,
    awayTeam :: TeamData,
    shirtNumbers :: ShirtNumbers
}

data TeamData = TeamData {
    players :: [PlayerData]
}

data PlayerData = PlayerData {
    playerRef :: String,
    formationPosition :: String,
    shirtNumber :: ShirtNumber
}

parseMetaFile :: String -> IO Metadata
parseMetaFile filename = do
    root <- Xml.loadXmlFromFile filename
    return $ makeMetadata root

makeMetadata :: Element -> Metadata
makeMetadata element =
    Metadata
      { homeTeam = parseTeamData homeTeamDataElement
      , awayTeam = parseTeamData awayTeamDataElement
      , shirtNumbers = shirtNumbers
      }
    where
    shirtNumbers = Map.union (collectShirtNumbers Tcb.Home homeTeam) (collectShirtNumbers Tcb.Away awayTeam)

    collectShirtNumbers teamKind teamData =
        Map.fromList $ map createPair (players teamData)
        where
        createPair playerData =
            -- Player data looks like: <MatchPlayer Formation_Place="1" PlayerRef="p167040" Position="Goalkeeper" ShirtNumber="1" Status="Start" />
            -- So for some reason the playerRef is in the format `p<player_id>` where the `player_id` part matches up with those given in the Event data
            -- in the actual f24 file. So to parse it we just drop the 'p' and read the rest as an integer.
            -- This player_id part we map to the shirt number, since this should correspond with the shirt-number used in the Tracab file.
            -- So basically here, we are going from the above MatchPlayer element to `(167040, 1)`.
            ( read $ drop 1 $ playerRef playerData
            , (teamKind, shirtNumber playerData)
            )


    homeTeam = parseTeamData homeTeamDataElement
    awayTeam = parseTeamData awayTeamDataElement
    homeTeamDataElement =
        head $ filter (hasAttributeWithValue "Side" "Home") allTeamData

    awayTeamDataElement =
        head $ filter (hasAttributeWithValue "Side" "Away") allTeamData

    allTeamData =
        Xml.getChildrenWithQName "TeamData" matchData
        where
        soccerFeed = element
        soccerDocument = last $ Xml.getChildrenWithQName "SoccerDocument" soccerFeed
        matchData = head $ Xml.getChildrenWithQName "MatchData" soccerDocument

    parseTeamData teamElement =
        TeamData { players = map parsePlayerData $ Xml.getChildrenWithQName "MatchPlayer" lineupElement }
        where
        lineupElement = head $ Xml.getChildrenWithQName "PlayerLineUp" teamElement

    parsePlayerData playerElement =
        PlayerData
            { playerRef = Xml.attrLookupStrict playerElement id "PlayerRef"
            , formationPosition = Xml.attrLookupStrict playerElement id "Position"
            , shirtNumber = Xml.attrLookupStrict playerElement read "ShirtNumber"
            }
